### ISSUE TITLE
Issues with AutoAugment

### ISSUE THREAD

**Issue by rwightman** [tensorflow/tpu#495]
I've been trying to replicate AutoAugment (currently for classification) based on the implementation here for EfficientNet. I've run into some issues and have a question.

1. Posterize as setup with the 'v0' policy for classification will result in black images for one of the sub policies. I can't see how that is intended or desired. Was this implementation as is used for training the EfficientNet models or was it moved from somewhere else with potential mistakes?

You've impl posterize as equivalent to PIL where the `bits` argument is the number of (most significant) bits to keep:
```
def posterize(image, bits):
  """Equivalent of PIL Posterize."""
  shift = 8 - bits
  return tf.bitwise.left_shift(tf.bitwise.right_shift(image, shift), shift)
```
The level to arg lamba for posterize is: `'Posterize': lambda level: (int((level/_MAX_LEVEL) * 4),),`

One of the sub-policies involving posterize is: `[('Posterize', 0.8, 2), ('Solarize', 0.6, 10)],`

`int((2/10)*4) = 0` and thus this posterize op keeps 0 bits and results in a black image. Also for your TF impl, shift an 8-bit value by 8 bits is potentially undefined (does it result in a no-op or shift by 7 on a TPU?). The PIL impl uses a LUT and just zeros the image as told.

Was this the intent or was something lost in translation? The paper states that Posterize will `Reduce the number of bits for each pixel to magnitude bits.  Range [4,8]`  

The above is a bit vague but suggests that it will leave the 4-8 MSB bits (so cut up the 4 LSB to unchanged at 8). This is not the current impl. Was the posterize function intended to be `shift=4-bits` or the level function be `lambda level: (int((level/_MAX_LEVEL) * 4) + 4,)` ?

2. Other question. Where did the V0 policy come from? It's not in the original paper and quite different from the ImageNet policy there. It uses SolarizeAdd which is only in the newer Object detection paper. But that obj detection paper does not include an Imagenet policy?

**Comment by rwightman** [comment]
Also, the AutoAugment as implemented here for classification is done in addition to the random crop + flip that is standard in ImageNet preprocessing. Is that the intent?

**Comment by BarretZoph** [comment]
Hello thank you for the questions!

This implementation was used for training the EfficientNet models and the black images were just a feature of what the autoaugment policy found to be useful. Sometimes having extremely noisy or even black images during the training would not hamper and even slightly help performance in certain cases. 

You are correct that this does shift the bits to 0 to 4 and not 4 to 8, which was not intended when this was first done. But this was what was used when the search procedure happened, which allows for the search to work around these issues since it has the ability to learn what ops and magnitudes to select.

The v0 policy was another policy that we found to work well, but not as well as the one in the original AutoAugment paper that was used for the results. This specific policy appeared to work better for the EfficientNet models. The new op was added in some searches right after we finished the paper.
